doc/man/man1/oconv.1
.\" RCSid "$Id$"
.TH OCONV 1 8/15/95 RADIANCE
.SH NAME
oconv - create an octree from a RADIANCE scene description
.SH SYNOPSIS
.B oconv
[
.B "\-i octree"
|
.B "\-b xmin ymin zmin size
][
.B "\-n objlim"
][
.B "\-r maxres"
][
.B \-f
][
.B \-w
][
-
]
[
.B "input .."
]
.SH DESCRIPTION
.I Oconv
adds each scene description
.I input
to
.I octree
and sends the result to the standard output.
Each
.I input
can be either a file name, or a command (enclosed in quotes
and preceded by a `!').
Similarly, the octree input may be given as a command preceded
by a `!'.
If any of the surfaces will not fit in
.I octree,
an error message is printed and the program aborts.
If no
.I octree
is given, a new one is created large enough for
all of the surfaces.
.PP
The
.I \-b
option allows the user to give a bounding cube for the
scene, starting at
.I "xmin ymin zmin"
and having a side length
.I size.
If the cube does not contain all of the surfaces, an error results.
The
.I \-b
and
.I \-i
options are mutually exclusive.
.PP
The
.I \-n
option specifies the maximum surface set size for
each voxel.
Larger numbers result in quicker octree generation,
but potentially slower rendering.
Smaller values may or may not produce faster renderings,
since the default number (6) is close to optimal for most scenes.
.PP
The
.I \-r
option specifies the maximum octree resolution.
This should be greater than or equal to the ratio of the largest
and smallest dimensions in the scene (ie. surface size or distance between 
surfaces).
The default is 16384.
.PP
The
.I \-f
option produces a frozen octree containing all the scene information.
Normally, only a reference to the scene files is stored in the
octree, and changes to those files may invalidate the result.
The freeze option is useful when the octree file's
integrity and loading speed is more important than its size, or
when the octree is to be relocated to another directory, and is
especially useful for creating library objects for the "instance"
primitive type.
If the input octree is frozen, the output will be also.
.PP
The
.I \-w
option suppresses warnings.
.PP
A hyphen by itself ('-') tells
.I oconv
to read scene data from its standard input.
This also implies the
.I \-f
option.
.PP
The only scene file changes that do not require octree regeneration are
modifications to non-surface parameters.
If the coordinates of a surface are changed, or any primitives are added
or deleted, oconv must be run again.
Programs will abort with a "stale octree" message
if they detect any dangerous inconsistencies between
the octree and the input files.
.PP
Although the octree file format is binary, it is meant to be portable
between machines.
The only limitation is that machines with radically different integer
sizes will not work together.
For the best results, the \-f option should be used if an octree is
to be used in different environments.
.SH DIAGNOSTICS
There are four basic error types reported by oconv:
.IP
warning - a non-fatal input-related error
.IP
fatal - an unrecoverable input-related error
.IP
system - a system-related error
.IP
internal - a fatal error related to program limitations
.IP
consistency - a program-caused error
.PP
Most errors are self-explanatory.
However, the following internal errors should be mentioned:
.IP "Too many scene files"
Reduce the number of scene files by combining them or using calls
to xform(1) within files to create a hierarchy.
.IP "Hash table overflow in fullnode"
This error is caused by too many surfaces.
If it is possible to create an octree for the scene at all,
it will have to be done in stages using the
.I \-i
option.
.SH EXAMPLE
To add book1, book2 and a
transformed book3 to the octree ``scene.oct'':
.IP "" .2i
oconv \-i scene.oct book1 book2 '\\!xform \-rz 30 book3' > newscene.oct
.SH AUTHOR
Greg Ward
.SH NOTES
In the octree, the names of the scene files are stored rather
than the scene information.
This means that a new octree must be generated whenever the
scene files are changed or moved.
Also, an octree that has been moved to a new directory will not be able to find
scene files with relative pathnames.
The freeze option avoids these problems.
.I make(1)
or
.I rad(1)
can be used to automate octree creation and maintenance.
.PP
The infamous "set overflow" error has been changed to a warning,
which allows imperfect octrees to be built with small holes
where too many polygons or other surfaces are crowded together
into a small space.
.SH "SEE ALSO"
getbbox(1), getinfo(1), make(1), obj2mesh(1), rad(1),
rpict(1), rvu(1), rtrace(1), xform(1)


src/ot/oconv.c
#ifndef lint
static const char RCSid[] = "$Id$";
#endif
/*
 *  oconv.c - main program for object to octree conversion.
 *
 *     7/29/85
 */

#include  "platform.h"
#include  "standard.h"
#include  "octree.h"
#include  "object.h"
#include  "otypes.h"
#include  "resolu.h"
#include  "oconv.h"

#define	 OMARGIN	(10*FTINY)	/* margin around global cube */

#define	 MAXOBJFIL	255		/* maximum number of scene files */

int  nowarn = 0;			/* supress warnings? */

int  objlim = 6;			/* # of objects before split */

int  resolu = 16384;			/* octree resolution limit */

CUBE  thescene = {{0.0, 0.0, 0.0}, 0.0, EMPTY};		/* our scene */

char  *ofname[MAXOBJFIL+1];		/* object file names */
int  nfiles = 0;			/* number of object files */

double	mincusize;			/* minimum cube size from resolu */

void  (*addobjnotify[])() = {NULL};	/* new object notifier functions */

static void addobject(CUBE  *cu, OBJECT	obj);
static void add2full(CUBE  *cu, OBJECT	obj, int  inc);


int
main(		/* convert object files to an octree */
	int  argc,
	char  *argv[]
)
{
	FVECT  bbmin, bbmax;
	char  *infile = NULL;
	int  inpfrozen = 0;
	int  outflags = IO_ALL;
	OBJECT	startobj;
	int  i;

	fixargv0(argv[0]);		/* sets global progname */

	ot_initotypes();

	for (i = 1; i < argc && argv[i][0] == '-'; i++)
		switch (argv[i][1]) {
		case '\0':				/* scene from stdin */
			goto breakopt;
		case 'i':				/* input octree */
			infile = argv[++i];
			break;
		case 'b':				/* bounding cube */
			thescene.cuorg[0] = atof(argv[++i]) - OMARGIN;
			thescene.cuorg[1] = atof(argv[++i]) - OMARGIN;
			thescene.cuorg[2] = atof(argv[++i]) - OMARGIN;
			thescene.cusize = atof(argv[++i]) + 2*OMARGIN;
			break;
		case 'n':				/* set limit */
			objlim = atoi(argv[++i]);
			break;
		case 'r':				/* resolution limit */
			resolu = atoi(argv[++i]);
			break;
		case 'f':				/* freeze octree */
			outflags &= ~IO_FILES;
			break;
		case 'w':				/* supress warnings */
			nowarn = 1;
			break;
		default:
			sprintf(errmsg, "unknown option: '%s'", argv[i]);
			error(USER, errmsg);
			break;
		}
breakopt:
	SET_FILE_BINARY(stdout);
	if (infile != NULL) {		/* get old octree & objects */
		if (thescene.cusize > FTINY)
			error(USER, "only one of '-b' or '-i'");
		nfiles = readoct(infile, IO_ALL, &thescene, ofname);
		if (nfiles == 0)
			inpfrozen++;
	} else
		newheader("RADIANCE", stdout);	/* new binary file header */
	printargs(argc, argv, stdout);
	fputformat(OCTFMT, stdout);
	printf("\n");

	startobj = nobjects;		/* previous objects already converted */

	for ( ; i < argc; i++)		/* read new scene descriptions */
		if (!strcmp(argv[i], "-")) {	/* from stdin */
			readobj(NULL);
			outflags &= ~IO_FILES;
		} else {			/* from file */
			if (nfiles >= MAXOBJFIL)
				error(INTERNAL, "too many scene files");
			readobj(ofname[nfiles++] = argv[i]);
		}

	ofname[nfiles] = NULL;

	if (inpfrozen && outflags & IO_FILES) {
		error(WARNING, "frozen octree");
		outflags &= ~IO_FILES;
	}
						/* find bounding box */
	bbmin[0] = bbmin[1] = bbmin[2] = FHUGE;
	bbmax[0] = bbmax[1] = bbmax[2] = -FHUGE;
	for (i = startobj; i < nobjects; i++)
		add2bbox(objptr(i), bbmin, bbmax);
						/* set/check cube */
	if (thescene.cusize == 0.0) {
		if (bbmin[0] <= bbmax[0]) {
			for (i = 0; i < 3; i++) {
				bbmin[i] -= OMARGIN;
				bbmax[i] += OMARGIN;
			}
			for (i = 0; i < 3; i++)
				if (bbmax[i] - bbmin[i] > thescene.cusize)
					thescene.cusize = bbmax[i] - bbmin[i];
			for (i = 0; i < 3; i++)
				thescene.cuorg[i] =
					(bbmax[i]+bbmin[i]-thescene.cusize)*.5;
		}
	} else {
		for (i = 0; i < 3; i++)
			if (bbmin[i] < thescene.cuorg[i] ||
				bbmax[i] > thescene.cuorg[i] + thescene.cusize)
				error(USER, "boundary does not encompass scene");
	}

	mincusize = thescene.cusize / resolu - FTINY;

	for (i = startobj; i < nobjects; i++)		/* add new objects */
		addobject(&thescene, i);

	thescene.cutree = combine(thescene.cutree);	/* optimize */

	writeoct(outflags, &thescene, ofname);	/* write structures to stdout */

	quit(0);
	return 0; /* pro forma return */
}


void
quit(				/* exit program */
	int  code
)
{
	exit(code);
}


void
cputs(void)					/* interactive error */
{
	/* referenced, but not used */
}


void
wputs(				/* warning message */
	const char  *s
)
{
	if (!nowarn)
		eputs(s);
}


void
eputs(				/* put string to stderr */
	const char  *s
)
{
	static int  inln = 0;

	if (!inln++) {
		fputs(progname, stderr);
		fputs(": ", stderr);
	}
	fputs(s, stderr);
	if (*s && s[strlen(s)-1] == '\n')
		inln = 0;
}

				/* conflicting def's in param.h */
#undef  tstbit
#undef  setbit
#undef  clrbit
#undef  tglbit

#define	 bitop(f,i,op)		(f[((i)>>3)] op (1<<((i)&7)))
#define	 tstbit(f,i)		bitop(f,i,&)
#define	 setbit(f,i)		bitop(f,i,|=)
#define	 clrbit(f,i)		bitop(f,i,&=~)
#define	 tglbit(f,i)		bitop(f,i,^=)


static void
addobject(			/* add an object to a cube */
	CUBE  *cu,
	OBJECT	obj
)
{
	int  inc;

	inc = (*ofun[objptr(obj)->otype].funp)(objptr(obj), cu);

	if (inc == O_MISS)
		return;				/* no intersection */

	if (istree(cu->cutree)) {
		CUBE  cukid;			/* do children */
		int  i, j;
		cukid.cusize = cu->cusize * 0.5;
		for (i = 0; i < 8; i++) {
			cukid.cutree = octkid(cu->cutree, i);
			for (j = 0; j < 3; j++) {
				cukid.cuorg[j] = cu->cuorg[j];
				if ((1<<j) & i)
					cukid.cuorg[j] += cukid.cusize;
			}
			addobject(&cukid, obj);
			octkid(cu->cutree, i) = cukid.cutree;
		}
		return;
	}
	if (isempty(cu->cutree)) {
		OBJECT  oset[2];		/* singular set */
		oset[0] = 1; oset[1] = obj;
		cu->cutree = fullnode(oset);
		return;
	}
					/* add to full node */
	add2full(cu, obj, inc);
}


static void
add2full(			/* add object to full node */
	CUBE  *cu,
	OBJECT	obj,
	int  inc
)
{
	OCTREE	ot;
	OBJECT	oset[MAXSET+1];
	CUBE  cukid;
	unsigned char  inflg[(MAXSET+7)/8], volflg[(MAXSET+7)/8];
	int  i, j;

	objset(oset, cu->cutree);
	cukid.cusize = cu->cusize * 0.5;

	if (inc==O_IN || oset[0] < objlim || cukid.cusize < mincusize) {
		if (oset[0] >= MAXSET) {
			sprintf(errmsg, "set overflow in add2full (%s)",
					objptr(obj)->oname);
			error(WARNING, errmsg);
			return;			/* XXX leaves a tiny hole */
		}				/* else add to set */
		insertelem(oset, obj);
		cu->cutree = fullnode(oset);
		return;
	}
					/* subdivide cube */
	if ((ot = octalloc()) == EMPTY)
		error(SYSTEM, "out of octree space");
					/* mark volumes */
	j = (oset[0]+7)>>3;
	while (j--)
		volflg[j] = inflg[j] = 0;
	for (j = 1; j <= oset[0]; j++)
		if (isvolume(objptr(oset[j])->otype)) {
			setbit(volflg,j-1);
			if ((*ofun[objptr(oset[j])->otype].funp)
					(objptr(oset[j]), cu) == O_IN)
				setbit(inflg,j-1);
		}
					/* assign subcubes */
	for (i = 0; i < 8; i++) {
		cukid.cutree = EMPTY;
		for (j = 0; j < 3; j++) {
			cukid.cuorg[j] = cu->cuorg[j];
			if ((1<<j) & i)
				cukid.cuorg[j] += cukid.cusize;
		}
					/* surfaces first */
		for (j = 1; j <= oset[0]; j++)
			if (!tstbit(volflg,j-1))
				addobject(&cukid, oset[j]);
					/* then this object */
		addobject(&cukid, obj);
					/* then partial volumes */
		for (j = 1; j <= oset[0]; j++)
			if (tstbit(volflg,j-1) &&
					!tstbit(inflg,j-1))
				addobject(&cukid, oset[j]);
					/* full volumes last */
		for (j = 1; j <= oset[0]; j++)
			if (tstbit(inflg,j-1))
				addobject(&cukid, oset[j]);
					/* returned node */
		octkid(ot, i) = cukid.cutree;
	}
	cu->cutree = ot;
}
